File: C++/代码随想录/Linked_List/19. Remove Nth Node From End of List.py
from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        dummy = ListNode(0)
        dummy.next = head
        first = second = dummy
        for _ in range(n + 1):
            first = first.next

        while first:
            first = first.next
            second = second.next

        second.next = second.next.next
        return dummy.next

# 测试用例
def printList(head):
    while head:
        print(head.val, end=" -> ")
        head = head.next
    print("None")

sol = Solution()

# 创建链表 1 -> 2 -> 3 -> 4 -> 5
head = ListNode(1, ListNode(2, ListNode(3, ListNode(4, ListNode(5)))))

# 删除倒数第二个节点
printList(sol.removeNthFromEnd(head, 2))  # 应该输出 1 -> 2 -> 3 -> 5 -> None

File: C++/代码随想录/Linked_List/19. Remove Nth Node From End of List.cpp
// 19. Remove Nth Node From End of List
// https://leetcode.com/problems/remove-nth-node-from-end-of-list/

#include "LinkedListUtils.hpp"

ListNode* removeNthFromEnd(ListNode* head, int n) {
    ListNode* dummy = new ListNode(0);
    dummy->next = head;

    ListNode* fast = dummy;
    ListNode* slow = dummy;

    // 先将 fast 指针移动 n+1 步
    for (int i = 0; i <= n; i++) {
        fast = fast->next;
    }

    // 同时移动 fast 和 slow 指针，直到 fast 到达链表末尾
    while (fast != nullptr) {
        fast = fast->next;
        slow = slow->next;
    }

    // 删除 slow 指针后的节点
    ListNode* toDelete = slow->next;
    slow->next = slow->next->next;
    delete toDelete;

    return dummy->next;
}

int main() {
    std::vector<int> values = {1, 2, 3, 4, 5, 6, 7};
    ListNode* head = createLinkedList(values);
    ListNode* removed = removeNthFromEnd(head, 2);
    printList(removed);

    return 0;
}
